src/components/TodoItem.tsx

```diff
@@ -0,0 +1,207 @@
+    toggledTodos,
+    setToggledTodos,
+    isToggleAllClicked,
+    isEveryTodoCompleted,
+    inputRef,
+  } = useContext(TodoContext);
+
+  const [
+    deletingTodo,
+    setDeletingTodo,
+  ] = useState<Todo | undefined>(undefined);
+
+  const [updatingTodo, setUpdatingTodo] = useState<number | null>(null);
+  const [updatedTitle, setUpdatedTitle] = useState(todo.title);
+  const [isToggled, setIsToggled] = useState(false);
+
+  const todoInputRef = useRef<HTMLInputElement>(null);
+
+  const isEqualTitle = updatedTitle.trim() === todo.title.trim();
+
+  useEffect(() => {
+    if (updatingTodo && todoInputRef.current) {
+      todoInputRef.current.focus();
+    }
+  }, [updatingTodo]);
+
+  const handleDoubleClick = (todoId: number) => {
+    setUpdatingTodo(todoId);
+  };
+
+  const handleTodoToggle = useCallback((todoId: number) => {
+    setToggledTodos(todoId);
+    setIsToggled(true);
+
+    updateTodo(todoId, { completed: !todo.completed })
+      .then(() => {
+        const updatedTodos = todos.map((t) => {
+          if (t.id === todoId) {
+            return { ...t, completed: !t.completed };
+          }
+
+          return t;
+        });
+
+        setToggledTodos(null);
+        setTodos(updatedTodos);
+      })
+      .catch(() => setError(ErrorMessage.UpdateTodo))
+      .finally(() => setIsToggled(false));
+  }, [setError,
+    setTodos,
+    todo.completed,
+    setToggledTodos,
+    todos]);
+
+  const handleDelete = (todoId: number) => {
+    removeTodo(todoId)
+      .then(() => {
+        setTodos(todos.filter(t => t.id !== todoId));
+      })
+      .catch(() => setError(ErrorMessage.DeleteTodo))
+      .finally(() => {
+        setDeletingTodo(undefined);
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      });
+
+    setDeletingTodo(todos.find(t => t.id === todoId));
+  };
+
+  const handleBlur = (todoId: number) => {
+    if (updatedTitle.trim() === '') {
+      handleDelete(todoId);
+
+      return;
+    }
+
+    setIsUpdating(true);
+
+    updateTodo(todoId, { title: updatedTitle })
+      .then(() => {
+        const updatedTodos = todos.map((t) => {
+          if (t.id === todoId) {
+            return { ...t, title: updatedTitle.trim() };
+          }
+
+          return t;
+        });
+
+        setUpdatingTodo(null);
+        setTodos(updatedTodos);
+      })
+      .catch(() => setError(ErrorMessage.UpdateTodo))
+      .finally(() => setIsUpdating(false));
+  };
+
+  const handleKeyUp = (
+    event: React.KeyboardEvent<HTMLInputElement>, todoId: number,
+  ) => {
+    if (event.key === Key.Enter && !isEqualTitle) {
+      handleBlur(todoId);
+    } else if (event.key === Key.Escape || isEqualTitle) {
+      setUpdatedTitle(todo.title);
+      setUpdatingTodo(null);
+      setIsUpdating(false);
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn('todo', { completed: todo.completed })}
+      onDoubleClick={() => handleDoubleClick(todo.id)}
```

Comment:
In this and the following examples, you do not need to directly pass the `todo.id` as an argument to your handle function. You can use `todo.id` in your functions directly.

Reply:
Ty, good point